### ISSUE TITLE
systemd - feature request? - a "WaitForExit" dependency

### ISSUE THREAD

**Issue by thx1111** [systemd/systemd#2616]
In a Service Unit which configures a wired and wireless bonded interface and runs wpa_supplicant on the wireless interface, and then DHCP clients on the bond0 interface, an explicit systemctl stop from within the defining service unit tears-down the entire group - DHCP clients, wpa_supplicant, and bond0.  This is done with
ExecStop=\
/usr/bin/systemctl stop dhcp6c@bond0.service ;\
/usr/bin/systemctl stop dhclient@bond0.service ;\
/usr/bin/wpa_cli -i %I terminate ;\

In this case a Requires dependency, as for instance "Requires=dhcp6c@bond0.service", does not work to create the correct sequencing.  The DHCP clients must _complete_ their IP address "Release"  and exit _before_ the wpa_supplicant process is terminated and the network link is actually lost.  Requires will initiate a termination of the DHCP clients, but will not wait before terminating the supplicant.  However, using an explicit systemctl stop does the right thing and waits for the DHCP clients to exit before terminating the supplicant.

Before and After do not work here.  The defining service unit which effectively starts the DHCP clients cannot be required to both run Before the DHCP clients are started and After the DHCP clients exit.  In practice, for instance, "Before=dhcp6c@bond0.service" has no effect on waiting for exit.

Is "systemctl stop" inside a service unit the best way to effect waiting for these other units to exit?  And the "proper" way to do this?

It would seem that, instead, something like an explicit WaitForExit in the [Unit] section would be a more formal mechanism, and would also allow systemd to recognize the dependency relationships.  The essential element of this dependency would be the idea that the defining service unit wait for the listed units to exit before executing ExecStop itself, in the same way that ExecStop processes themselves wait for one process to exit before executing the next process in sequence.

**Comment by poettering** [comment]
not sure I follow. But, as documented, in systemd the shutdown order of two units is always the reverse of the startup order. i.e. if your DHCP unit is After= your WPA unit, then this means two things:
1. The DHCP unit is started after the WPA unit
2. The WPA unit is stopped after the DHCP unit

Wich I think is what you are looking for, no?

Generally, I think using "systemctl stop" in ExecStop= or in fact any ExecXYZ= lines is primarily an indication for a misunderstanding.

**Comment by thx1111** [comment]
No, that does not work correctly.

Yes, in a sense, the WPA unit is "stopped" after the DHCP unit, for instance "systemd[1]: Stopping ISC dhclient on interface bond0...", BUT, systemd does not wait for the DHCP process to exit before "systemd[1]: Stopping bond0-wlp3s0-wpa_supplicant-nl80211...", which otherwise will _immediately_ execute "ExecStop=/usr/bin/wpa_cli -i %I terminate", which will _immediately_ terminate the wireless link, _before_ the DHCP client has negotiated an IP address "release".

 That is a "fail".

Any misunderstanding would have to do with what is meant by the idea of "reverse order" in the documentation.  Perhaps there is something, in effect, missing from the [Service] section of the "defining" or "called" unit.  I notice there are, for instance, Type=notify and Type=idle. 

Type=notify says that "it is expected that the daemon sends a notification message via sd_notify(3) or an equivalent call when it has finished starting up."  I assume that ISC dhclient does not have any built-in sd_notify(3) mechanism, and the documentation refers to "starting up", not "shutting down".

Type=idle documentation says "actual execution of the service binary is delayed until all jobs are dispatched".  Presumably "service binary" refers to the ExecStart process, but what is meant by "all jobs"?  The term "dispatched", again, seems to refer to "starting up" and not "shutting down".  And the described purpose has to do with output display, not with process ordering.

Yes, using "systemctl stop" in ExecStop= seems wrong, but it also "just works".  Maybe there needs to be a [Service] section "Type=waitforexit" included in the "called" unit?

Or should the "called" unit, dhclient, be configured as "Type=oneshot", "RemainAfterExit=yes" instead of "Type=simple"?  I am supposing that with Type=oneshot, systemd would no longer manage the ExecStart process if it were to "crash" or be "killed".  That would not be acceptable.

**Comment by poettering** [comment]
I have trouble understanding what you want to say, but yes, if there's a service foo.service and a service bar.service and bar.service has After=foo.service, and both are running, and then both are to be stopped, then systemd will first terminate bar.service, wait for it to fully shut down, and then begin terminating foo.service too.

**Comment by poettering** [comment]
I am not sure I groke at all what you are saying, but I am pretty sure this is just a misunderstanding how things work. Note that by default systemd will send SIGTERM to a service, wait for it to exit and then proceed with the next job (that declared an order against it). That it does for units without ExecStop= set. If you however set an ExecStop= it is expected that it is a tool that waits synchronously for the service to shut down on its own. If the ExecStop= tool of your choice doesn't do that however, does something completely unrelated and does not actually wait for the service to shut down, then things fall apart since that's not what ExecStop= is for.

**Comment by thx1111** [comment]
> and bar.service has After=foo.service, and both are running, and then both are to be stopped,
> then systemd will first terminate bar.service, wait for it to fully shut down, and then begin
> terminating foo.service too.

And, reading at man systemd.unit,  "Note that when two units with an ordering dependency between them are shut down, the inverse of the start-up order is applied. i.e. if a unit is configured with After= on another unit, the former is stopped before the latter if both are shut down."

Then, there is a bug.  That is _not_ what is happening.  I am actually, right now, running this, on systemd 229, though this is the same as on systemd 228.  The units of interest are

bond0-supplicant@.service      foo.service        "then begin terminating"
dhclient@.timer
dhclient@.service                     bar.service        "first terminate"

In the "dhclient@.service" Service Unit File, with "%i" being "bond0", there is
        After=bond0-supplicant@%i.service

This says that dhclient will _start after_ bond0-supplicant.  Conversely, when stopped, dhclient will "first terminate" _before_ bond0-supplicant.

To test this - again, but with exactly the form you have described - I have removed the previous
"ExecStop=/usr/bin/systemctl stop dhclient@bond0.service".

Then first, "systemctl stop bond0-supplicant@wlp3s0.service"
then "systemctl start bond0-supplicant@wlp3s0.service", and check that everything is in its proper state.

Then, "systemctl stop bond0-supplicant@wlp3s0.service & journalctl -afn100"
I am also watching, on the DHCP Server, "tcpdump -elnSvvvienp3s0 port 67 or port 68 or port 546 or port 547".

The result is
Feb 15 09:45:45 garnet systemd[1]: Stopping bond0-wlp3s0-wpa_supplicant-nl80211...
...
Feb 15 09:45:45 garnet wpa_cli[16229]: OK
Feb 15 09:45:45 garnet wpa_supplicant[16194]: wlp3s0: CTRL-EVENT-DISCONNECTED bssid=00:17:3f:3a:cd:7a reason=3 locally_generated=1
Feb 15 09:45:45 garnet kernel: bond0: link status definitely down for interface wlp3s0, disabling it
...
Feb 15 09:45:45 garnet wpa_supplicant[16194]: wlp3s0: CTRL-EVENT-TERMINATING
Feb 15 09:45:45 garnet kernel: bond0: Releasing backup interface wlp3s0
Feb 15 09:45:45 garnet kernel: bond0: Releasing active interface enp0s10
Feb 15 09:45:45 garnet dhclient[16198]: receive_packet failed on bond0: Network is down
...
Feb 15 09:45:45 garnet dhclient[16198]: receive_packet failed on bond0: Network is down
Feb 15 09:45:45 garnet kernel: bond0 (unregistering): Released all slaves
Feb 15 09:45:45 garnet systemd[1]: Stopped bond0-wlp3s0-wpa_supplicant-nl80211.
...
Feb 15 09:45:45 garnet systemd[1]: Stopping ISC dhclient on interface bond0...
Feb 15 09:45:45 garnet dhclient[16270]: Killed old client process
Feb 15 09:45:45 garnet dhclient[16270]: Killed old client process
Feb 15 09:45:46 garnet dhclient[16270]: No broadcast interfaces found - exiting.
Feb 15 09:45:46 garnet systemd[1]: Stopped ISC dhclient on interface bond0.

Of course, the DHCP Server never receives the release.

As you can see there, "Stopping ISC dhclient on interface bond0..." the stopping occurs _after_ "wpa_cli[16229]: OK" (/usr/bin/wpa_cli -i %I terminate) and _after_ "Stopped bond0-wlp3s0-wpa_supplicant-nl80211."  "bar.service" is not "first terminate".  "foo.service" is "first terminate".

And, to put this in perspective, for whatever it is worth, from direct experience, I know that using instead "ExecStop=/usr/bin/systemctl stop dhclient@bond0.service" does act differently, stopping dhclient _before_ stopping wpa_supplicant.  It is an effective "work-around", then, to what appears to be a bug.

There is a second problem with the proposed After=foo.service in the bar.service Service Unit file, and administrative issue, in contrast to using Before=bar.service in the foo.service Unit file.  I am assuming that these two configurations are meant to be equivalent.  Hopefully, they are equivalent.  "bar.service" is meant to be, and is, a "generic" Service file.  It does not need to "know" anything about "every other" Service file which might make reference to it.  Otherwise, the administrator would have to go around custom modifying every referenced .service file, to add After= referencing the "calling" Service Unit.  I will assume that this is not what you meant to imply.  And also, I have tried using "Before=dhclient@bond0.service" in the "bond0-supplicant@.service" file.  That also does not work.

And, to be clear, completely removing the use of dhclient@.timer make no difference to the operation described here.  The results are exactly the same.

**Comment by poettering** [comment]
It's really hard to follow what you are doing. If you want me to help you, could you please extract a minimal test case that reproduces the issue out of two minimal units actually called "foo.service" and "bar.service". Then let me know what this shows that you don't expect, and what you expect that this doesn't result in. 

Use /bin/sleep and /bin/echo as daemon processes for this test case.

**Comment by poettering** [comment]
BTW: are you aware that ordering and requirement deps are fully orthogonal in systemd? i.e. if you want that a unit A being stopped also shall result in a unit B being stopped, then B needs a Requires=A or so, to make that happen, and that's independent from the ordering between the operations. See systemd(1) about this.

**Comment by thx1111** [comment]
Are you saying that Before and After have no effect on a Stop ordering unless they are associated with a Requires, or presumably also a BindsTo?

I have supposed that a "RequiredBy=A" in unit B is equivalent to a "Requires=B" in unit A, and that an "After=A" in unit B is equivalent to a "Before=B" in unit A, and that a pairing of, for instance, "Requires" and "Before/After" are generally necessary, preferably in unit A, assuming that unit B is "generic".  In the example above, there was no "Requires".  Retesting with
Requires=dhclient@bond0.service
Before=dhclient@bond0.service
in bond0-supplicant@.service does not change the Stop ordering.

I also notice that I made a mistake above, using, effectively,
After=bond0-supplicant@bond0.service
instead of the correct
After=bond0-supplicant@wlp3s0.service
in the "dhclient@.service" file, in the example above.  Oddly, correcting this makes no difference in the result, suggesting that, whatever is actually going on, these Before and After directives are not being pulled into the start and stop ordering mechanism.

This is all pretty fundamental to systemd, so it seems I will have to build a simple test unit set, to illuminate the problem.  Still, please note that the above example is operating through the intermediary of the bond0 device unit, which is being created by an ip link command.  This is more complex than using simple processes like sleep and echo.  It is not clear that systemd has any way to anticipate the relationship between executing wpa_supplicant, and the "appearance" of the bond0 device, where dhclient is then activated by the appearance of bond0.  This may be the root of the problem, where "ip link ..." is embeded in an ExecStartPre rather than in ExecStart.  Otherwise, "ip link ..." would seem to have to be contained in a separate service file, with Type=oneshot and ExecStart=...ip link....  Even then, it is not clear to me how systemd would track the ordering of the ExecStop=...ip link..., which takes down bond0, and the separate ExecStop=...wpa_cli..., which takes down wpa_supplicant, and associate these with stopping the DHCP client.

Note here that, without the systemctl stop, dhclient only stops because of a "BindsTo=sys-subsystem-net-devices-%i.device", where %i is bond0.  How else is systemd suppose to know that stopping wpa_supplicant is suppose to stop dhclient?

Retesting with
Requires=sys-subsystem-net-devices-bond0.device
Before=sys-subsystem-net-devices-bond0.device
makes no difference and does not stop dhclient before wpa_supplicant.

For the sake of argument, what are the "proper" or "formal" service file components required to create the correct DHCP stop ordering?

Start:
1) bond0 -> 2) wpa_supplicant and dhclient, while noting that wpa_supplicant and bond0 are not dependent upon the DHCP client or clients, though there could be a Wants.  Here also, the sudden "appearance" of a wireless device can be used to activate the bond0 start sequence, but ignore that for the moment.

Stop:
1) dhclient -> 2) wpa_supplicant -> 3) bond0, while noting that the DHCP client or clients may or may not still exist at this point.  Here also, any "disappearance" of the bond0 device or the wireless device should stop and completely "tear-down" the group.

Again, this all works just fine with an "improper" systemctl stop in the service file.  Rather, the question is, what is the proper and formal systemd construction needed to achieve that same result, to create the correct Stop ordering chain?

**Comment by thx1111** [comment]
> If you however set an ExecStop= it is expected that it is a tool that waits synchronously
> for the service to shut down on its own. If the ExecStop= tool of your choice doesn't do that
> however, does something completely unrelated and does not actually wait for the service to
> shut down, then things fall apart since that's not what ExecStop= is for.

Aha!  Perhaps _that_ is the simple answer to this issue.  Let me ask a different question.  Effectively, what I am doing is _avoiding_ directly using
ExecStop=/usr/bin/dhclient -r -pf /run/dhclient-%i
or even
ExecStop=/usr/bin/dhclient -d -r -pf /run/dhclient-%i
with %i=bond0 in bond0-supplicant@wlp3s0.service.  When dhclient -d -r ... is used directly, it
it apparently exits immediately and continues running as a daemon in the background.  So, of course, then wpa_cli...terminate runs, taking down the network, while dhclient is still chatting to the server.

My naive work-around is to instead run
ExecStop= /usr/bin/systemctl stop dhclient@bond0.service
in bond0-supplicant@wlp3s0.service, which then calls
ExecStop=/usr/bin/dhclient -r -pf /run/dhclient-%i
with %I=bond0 in dhclient@bond0.service.  systemctl is well behaved and waits for dhclient to exit before returning to bond0-supplicant@wlp3s0.service.  So dhclient is finished before wpa_cli...terminate is run.

And so, the new question is, is that the "right" thing to do, to deal with dhclient?  Or?

**Comment by poettering** [comment]
Again, if you place "systemctl stop" invocations in any of the ExecXYZ= lines, you are doing something wrong. Use dependencies to bind and order runtime status of units together, don't use systemctl invocations.

Also, unless there's a really string reason to, never define ExecStop= at all. sytemd's internal SIGTERM logic should be fine for most cases.

**Comment by thx1111** [comment]
I cannot distinguish a systemd bug from a mistaken service unit configuration based upon generalities.  I need to know specifically what dependency components are necessary and sufficient to create a functional and effective "Stop ordering".

Are you saying that Before and After have no effect on a Stop ordering unless they are associated with a Requires, or presumably also, a BindsTo?  And that a Wants dependency is _not_ sufficient to effect a Stop ordering?

**Comment by poettering** [comment]
again, ordering and requirement dependencies are fully orthogonal (which is a fancy word for "independent"), which means After= doesn't pull in, and Wants= doesn't order, and that typically both are used in combination.

Please have a look at the docs of systemd. See the documentation for Wants=, After=, and systemd(1) itself. It's terse, but it describes all this.

**Comment by jrynyt** [comment]
Related but tangential question: I have a service that needs to start early at boot and stop early on shutdown. systemd After/Before relationships seem to make this impossible since order is reversed on shutdown.

One way is to create another oneshot service that stops in the right place and does `ExecStop=/usr/bin/systemctl stop other-service`, which I agree feels wrong, but it works.

`/usr/lib/systemd/system-shutdown/` is another option, but I'd rather keep all the start/stop scripts in the same place, and nobody will likely know to look there.

Any better ideas?

**Comment by gggeek** [comment]
Hello.
Sorry in advance if commenting on an old issue is frowned against in this project, but I seem to have stumbled on the same "problem".
My needs are: making sure that a custom service finishes execution cleanly before the system is shut down.
To that effect, I have written the following service definition, plus a couple of shell scripts:
```
[Unit]
Description=Service taking a few minutes for shutdown
After=local-fs.target remote-fs.target network.target
[Service]
Type=simple
ExecStart=/home/tests/systemd/parent.sh
TimeoutStopSec=310
[Install]
WantedBy=multi-user.target
```
what `parent.sh` does is: upon starts, it executes `child.sh &` and loops forever. On trapping TERM, it keeps checking if the child process is executing (via usage of commands such as `ps` and `sed`), and terminates with exit code 0 when child is not found any more. What `child.sh` does is: it loops forever. On trapping TERM, it waits for 3 minutes and then terminates with exit code 0. Both scripts echo a debug message every second or so to allow troubleshooting their status.

When I try to stop the service from the command line using `systemctl`, everything works perfectly: the parent waits for the process to stop for 3 minutes.

When I initiate a system shutdown, what happens is: the parent process gets sent the TERM signal and initiates shutdown, and at the same time (most likely a millisecond later), systemd starts shutting down both the network and filesystems. Then it waits for the `parent` service to finish (ie. process termination) before effectively stopping the OS.

But what I need is: the network and filesystem should not "begin to stop" until the custom service has "finished to stop".

I hope the above is clear.

Tested with systemd as found on Ubuntu 18.04.2 (237-3ubuntu10.21)

**Comment by gggeek** [comment]
PS: just thought that I'd add that I have tried to "do my homework", spending a considerable amount of time searching for the solution to the above problem and thinkering with different service definitions.

What I found is:
- a couple of issues like this one, all closed
- a lot of blog posts / stackoverflow answers, all mentioning slightly different solutions
- a few raging rants

Which leads me to think that this is still an area which could do with some improvement, either in docs or in tooling.

Example alternative solutions that do not seem to work:
- using scripts in /usr/lib/systemd/system-shutdown/ (as those get executed last at shutdown, not first)
- adding `Before=shutdown.target reboot.target halt.target` to the service definition

**Comment by arvidjaar** [comment]
> When I initiate a system shutdown, what happens is: the parent process gets sent the TERM signal and initiates shutdown, and at the same time (most likely a millisecond later), systemd starts shutting down both the network and filesystems.

Please provide full journalctl output with systemd log level set to debug. You can do it with `kill -RTMIN+22 1` before initiating shutdown; make sure persistent journal is enabled and get messages after reboot with `journalctl -b -1`.

**Comment by gggeek** [comment]
@arvidjaar thanks for suggesting to enable debug logging.
I am doing another round of testing, as I suspect that my original conclusions might have been flawed: I was in fact monitoring shutdown of the server in question via an ssh connection and a samba share. The fact that those do get stopped before the slow-to-die test service terminates are in fact not necessarily a sign that local disk has been unmounted or network stopped (doh). The journald log entries also might have confused me... I have now added a 'wget' call at the end of the termination routine of the child slow-to-die shell script and that one succeeds...
Next step is to test with proper executables instead of dummy shell scripts :-)
